src/robot/parsing/model/visitor.py

```diff
@@ -49,19 +88,53 @@ def visit_Statement(self, node):
             ...
     """
 
-    def visit(self, node: Node):
-        visitor = self._find_visitor(type(node)) or self.generic_visit
-        visitor(node)
+    def visit(self, node: Node) -> None:
+        visitor = self._find_visitor(type(node))
+        visitor(self, node)
+
+    def generic_visit(self, node: Node) -> None:
+        for value in _iter_field_values(node):
+            if value is None:
+                continue
+            if isinstance(value, list):
+                for item in value:
+                    self.visit(item)
+            else:
+                self.visit(value)
 
 
-class ModelTransformer(ast.NodeTransformer, VisitorFinder):
+class ModelTransformer(VisitorFinder["Node|list[Node]|None"]):
     """NodeTransformer that supports matching nodes based on their base classes.
 
     See :class:`ModelVisitor` for explanation how this is different compared
     to the standard `ast.NodeTransformer
     <https://docs.python.org/library/ast.html#ast.NodeTransformer>`__.
     """
 
-    def visit(self, node: Node):
-        visitor = self._find_visitor(type(node)) or self.generic_visit
-        return visitor(node)
+    def visit(self, node: Node) -> "Node|list[Node]|None":
+        visitor = self._find_visitor(type(node))
+        return visitor(self, node)
+
+    def generic_visit(self, node: Node) -> "Node|list[Node]|None":
+        for field, old_value in iter_fields(node):
+            if old_value is None:
+                continue
+            if isinstance(old_value, list):
+                new_values = []
+                for value in old_value:
+                    new_value = self.visit(value)
+                    if new_value is None:
+                        continue
+                    if isinstance(new_value, list):
+                        new_values.extend(new_value)
+                        continue
+                    new_values.append(new_value)
+                old_value[:] = new_values
+            else:
+                new_node = self.visit(old_value)
+                if new_node is None:
+                    delattr(node, field)
+                else:
+                    setattr(node, field, new_node)
```

Comment:
This looks rather complicated and thus error prone. I strongly believe this PR should concentrate on visitor method caching and possible other enhancements left for future.

Reply:
i will remove this for now, although it also brings a little more performance. And the way we use it here breaks the Liskov substitution principle

Reply:
We can also change it later if wanted. However, these are only slightly optimized forms from the original implementation of `NodeVisitor` and `NodeTransformer`.